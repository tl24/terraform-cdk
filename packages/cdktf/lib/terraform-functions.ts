import { Collection } from "./functions/collection";
import { Crypto } from "./functions/crypto";
import { DateTime } from "./functions/datetime";
import { Encoding } from "./functions/encoding";
import { Filesystem } from "./functions/filesystem";
import { Network } from "./functions/network";
import { Numeric } from "./functions/numeric";
import { String } from "./functions/string";
import { Type } from "./functions/type";

export { Collection } from "./functions/collection";
export { Crypto } from "./functions/crypto";
export { DateTime } from "./functions/datetime";
export { Encoding } from "./functions/encoding";
export { Filesystem } from "./functions/filesystem";
export { Network } from "./functions/network";
export { Numeric } from "./functions/numeric";
export { String } from "./functions/string";
export { Type } from "./functions/type";

export class Fn {
  public static collection = Collection;
  public static crypto = Crypto;
  public static dateTime = DateTime;
  public static encoding = Encoding;
  public static filesystem = Filesystem;
  public static network = Network;
  public static numeric = Numeric;
  public static str = String;
  public static type = Type;
}
