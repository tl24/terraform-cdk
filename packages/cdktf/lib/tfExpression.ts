import { IResolvable, IResolveContext } from "./tokens/resolvable";
import { Intrinsic } from "./tokens/private/intrinsic";
import { Tokenization } from "./tokens";

export class TFExpression extends Intrinsic {
  public isInnerTerraformExpression = false;
}

export class Reference extends TFExpression {
  constructor(private identifier: string) {
    super(identifier);
  }

  public resolve(): string {
    return this.isInnerTerraformExpression
      ? this.identifier
      : `\${${this.identifier}}`;
  }
}
export function ref(identifier: string): Reference {
  return new Reference(identifier);
}

export class FunctionCall extends TFExpression {
  constructor(private name: string, private args: Expression[]) {
    super({ name, args });

    this.args.forEach((arg) => {
      if (arg instanceof TFExpression) {
        arg.isInnerTerraformExpression = true;
      }
    });
  }

  public resolve(context: IResolveContext): string {
    const expr = `${this.name}(${this.args
      .map((arg) => {
        if (Tokenization.isResolvable(arg)) {
          return arg.resolve(context);
        }

        return typeof arg === "string" ? `"${arg}"` : arg;
      })
      .join(", ")})`;

    return this.isInnerTerraformExpression ? expr : `\${${expr}}`;
  }
}
export function call(name: string, args: Expression[]): FunctionCall {
  return new FunctionCall(name, args);
}

export type Expression =
  | Reference
  | FunctionCall
  | string
  | number
  | boolean
  | IResolvable;
