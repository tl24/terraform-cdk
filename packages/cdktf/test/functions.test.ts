import { Testing, TerraformStack, TerraformOutput, Fn } from "../lib";
import { TerraformVariable } from "../lib/terraform-variable";

test("static values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.Numeric.abs(-42),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("dynamic values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.Numeric.abs(variable.value),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("spreaded mixed values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.Numeric.max(10, variable.value, 200),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("spreaded token value", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "list(number)",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.Numeric.max(variable.value),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("string values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.Numeric.parseInt("-210", 10),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("mixed string spreads values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.String.format(
      "There are %d out of %d lights are on in %s",
      variable.value,
      4,
      "Hamburg"
    ),
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("combined functions", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const index = new TerraformVariable(stack, "index", {
    type: "number",
  });

  const list = new TerraformVariable(stack, "list", {
    type: `list(object({
      internal = number
      external = number
      protocol = string
    }))`,
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.Type.try(
      Fn.Collection.lookup(
        Fn.Collection.element(list.value, index.value),
        "internal",
        "waaat"
      ),
      Fn.DateTime.timestamp(),
      Fn.Crypto.uuid()
    ),
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});
