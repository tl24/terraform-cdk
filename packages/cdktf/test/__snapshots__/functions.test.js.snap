// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combined functions 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"variable\\": {
    \\"index\\": {
      \\"type\\": \\"number\\"
    },
    \\"list\\": {
      \\"type\\": \\"list(object({\\\\n      internal = number\\\\n      external = number\\\\n      protocol = string\\\\n    }))\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${try(lookup(element(var.list, var.index), \\\\\\"internal\\\\\\", \\\\\\"waaat\\\\\\")\, timestamp(), uuid())}\\"
    }
  }
}"
`;

exports[`dynamic values 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"variable\\": {
    \\"test-var\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${abs(var.test-var)}\\"
    }
  }
}"
`;

exports[`mixed string spreads values 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"variable\\": {
    \\"test-var\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${format(\\\\\\"There are %d out of %d lights are on in %s\\\\\\", var.test-var, 4, \\\\\\"Hamburg\\\\\\")}\\"
    }
  }
}"
`;

exports[`spreaded mixed values 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"variable\\": {
    \\"test-var\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${max(10, var.test-var, 200)}\\"
    }
  }
}"
`;

exports[`spreaded token value 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"variable\\": {
    \\"test-var\\": {
      \\"type\\": \\"list(number)\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${max(var.test-var)}\\"
    }
  }
}"
`;

exports[`static values 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${abs(-42)}\\"
    }
  }
}"
`;

exports[`string values 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"output\\": {
    \\"test-output\\": {
      \\"value\\": \\"\${parseInt(\\\\\\"-210\\\\\\", 10)}\\"
    }
  }
}"
`;
